Added support for end2end testing using easy-server files

andy-maier opened the pull request:
See commit message.

Particular review points are:
* The data structure in the es_server.yml and es_vault.yml files (see comments in those files).
* The concept of WBEM server functions in the es_server.yml files and how it is used in test functions to skip the tests.
* The specific WBEM server functions that are currently defined (see es_server.yml file).

Note: The validation of the user-defined sections of the es_server.yml and es_vault.yml files with a JSON schema needs to be in a future separate issue, because that requires an extension in the underlying pytest-easy-server package.

KSchopmeyer commented on tests/end2endtest/es_server.yml:
```diff
@@ -0,0 +1,87 @@
+#                                  # created pywbem.WBEMConnection object.
+#                                  # Optional, default: null.
+#
+#       functions:                 # List of functions implemented by the
```
The word function confused me in particular because it conflicts with programming use of the word function.  I suggest either functionality or capability or even service.  Even "server process" might be logical.

andy-maier replied:
DONE - changed to "capabilities".

KSchopmeyer commented on tests/end2endtest/es_server.yml:
```diff
@@ -0,0 +1,87 @@
+# server_groups:                   # WBEM server groups.
+#                                  # Optional, default: no groups defined.
+#
+#   NICKNAME42:                    # Nickname of WBEM server group.
```
Not really a big issue but why do we call it nickname rather than name???

andy-maier replied:
That would be a change in easy-server, since it is part of the stuff inherited from there. Also, I wanted to have a term that is sufficiently far away from the "hostname" of the server and makes it clear that it only exists in the realm of easy-server.

KSchopmeyer commented on tests/end2endtest/es_server.yml:
```diff
@@ -0,0 +1,87 @@
+#                                  # Docker image.
+#                                  # Optional, default: null.
+#
+#       docker_port_mapping:       # Port mapping for Docker image.
```
The port mapping confuses me:

1. What is straight mapping? I assume it is same port on both host and docker.
2. How do you map multiple ports (ex 5988 and 5989)

andy-maier replied:
1. correct.
2. you cannot. Each server has a single URL, so you need to decide which protocol and port you use. If you want to test both protocols and ports, define two servers.

KSchopmeyer commented on tests/end2endtest/es_server.yml:
```diff
@@ -0,0 +1,87 @@
+#         - interop                # Has an Interop namespace
+#         - namespaces             # Supports management of namespaces
+#         - profiles               # General support for mgmt profiles
+#         - smis                   # Support for SMI-S profiles
```
Not sure if smis is good example.  I could see specific smis profiles however. Some other possibilities include slp,  ssl (i.e. for testing ssl capabilities), subscriptions/indications, pull operations,

andy-maier replied:
These functions are not examples, but actually existing end2end tests.

For SMI-S specific tests, we have two test modules test_profile_snia_smis.py that depends on the SMI-.S way of advertising profiles and test_profile_snia_server.py that depends on the SNIA Server profile. So we could split "smis" into two functions, but I think that every SMI-S implementation will support the SNIA Server profile, so you would never find one without the other. Anyway, we could still split "smis" into something like "snia-smis" and "snia-server".

andy-maier replied:
DONE - I have split it into "snia-smis" and "snia-server".

KSchopmeyer reviewed (commented):
I made several comments

andy-maier commented:
Updated the PR to address your comments. Please review again.

KSchopmeyer commented on docs/changes.rst:
```diff
@@ -265,6 +265,14 @@ Released: not yet
   to remove a ResourceWarning about unclosed sockets. (issue #2610)
 
+* Test: Added support for end2end testing of WBEM servers based on
+  easy-server server and vault files. The server files can specify WBEM servers
```
suggest change easy-server to easy-server python package

andy-maier replied:
DONE - changed to "server and vault files of the 'easy-server' Python package".

KSchopmeyer reviewed (commented):
Made a couple of minor comments and a note that apparently docker allows mapping not only single ports but ranges of ports (something like 5988-5989:9988-9989 although I have not used that capability.

andy-maier commented on tests/end2endtest/es_server.yml:
```diff
@@ -0,0 +1,88 @@
+#         image: PORT              # Port number provided by Docker image.
+#                                  # Mandatory.
+#
+#         local: PORT              # Port number used on local system.
```
Yes, it does. Changed to "host". DONE

andy-maier commented:
Added a note to the description of the port mapping to address the comment about mapping multiple ports:
```
#                                  # Note that each server defines the single
#                                  # port that is to be used for testing, so if
#                                  # you want to use both the HTTP and HTTPS
#                                  # ports, define a server for each.
```

andy-maier commented:
I added a second pegasus server for the second HTTP ports. That drove some further changes:
* Fixed the incorrect description of server groups in the es_server.yml file
* Made the docker container name unique by prepending the server nickname

Also, addressed the DeprecationWarning about using yaml.load().
